File: packages/js/src/from-kymojson.ts
/**
 * `.kymo.json` loader — the kymo.json interchange format → a resolved `Diagram`.
 *
 * The inverse of `toKymoJson` (`to-kymojson.ts`; KYMOJSON-MAP-001), and the JS mirror
 * of `packages/python/src/kymo/from_kymojson.py`. A `.kymo.json` file already holds a
 * fully-resolved model (positions baked in), so callers render it directly with no
 * layout / alignment pass. Unknown top-level fields are ignored (forward compatibility).
 */
import {
  makeComponent, makeDiagram, makeEdge, makeRegion,
  type Diagram, type Point, type Shape, type Side,
} from "./model.js";
import type { LayoutNode } from "./layout.js";

export const FORMAT = "kymo.json";

type Obj = Record<string, any>; // eslint-disable-line @typescript-eslint/no-explicit-any

function layoutNode(n: Obj): LayoutNode {
  if (n.t === "id") return { t: "id", id: n.id };
  return { t: "group", dir: n.dir, children: (n.children as Obj[]).map(layoutNode) };
}

/** Load a `.kymo.json` string into a fully-resolved `Diagram`. */
export function parseKymoJson(text: string): Diagram {
  const payload = JSON.parse(text);
  if (payload.format !== FORMAT) {
    throw new Error(`not a kymo.json document (format=${JSON.stringify(payload.format)}, expected ${JSON.stringify(FORMAT)})`);
  }
  const d = payload.diagram as Obj;
  return makeDiagram({
    width: d.width, height: d.height, title: d.title, subtitle: d.subtitle,
    components: (d.components as Obj[]).map((c) => makeComponent({
      id: c.id, name: c.name, subtitle: c.subtitle, icon: c.icon, shape: c.shape as Shape,
      accent: c.accent, pos: c.pos as Point, size: c.size as Point | null, parent: c.parent,
      align: c.align, alignGap: c.align_gap, alignOffset: c.align_offset as Point,
      labelBox: c.label_box ?? null,
    })),
    regions: (d.regions as Obj[]).map((r) => makeRegion({
      id: r.id, label: r.label, bounds: r.bounds, contains: r.contains,
      padding: r.padding, paddingBottom: r.padding_bottom, style: r.style, icon: r.icon,
      layout: r.layout, pos: r.pos as Point | null, gap: r.gap, align: r.align,
      visible: r.visible, borderDash: r.border_dash, borderStroke: r.border_stroke,
      labelAnchor: r.label_anchor, labelPosition: r.label_position,
    })),
    edges: (d.edges as Obj[]).map((e) => makeEdge({
      src: e.src, dst: e.dst, label: e.label, style: e.style,
      srcAnchor: e.src_anchor as Side | null, dstAnchor: e.dst_anchor as Side | null,
      route: e.route, via: e.via as Point[], srcOffset: e.src_offset as Point,
      dstOffset: e.dst_offset as Point, labelOffset: e.label_offset as Point,
      labelAnchor: e.label_anchor, labelSmall: e.label_small, labelPos: e.label_pos as Point | null,
      dashed: e.dashed, noArrow: e.no_arrow, trunkOffset: e.trunk_offset,
      sharedPort: e.shared_port, points: e.points as Point[] | null, bpmnFlow: e.bpmn_flow,
    })),
    layoutTrees: ((d.layout_trees ?? []) as Obj[]).map(layoutNode),
  });
}

File: packages/python/src/kymo/from_kymojson.py
"""`.kymo.json` loader — the kymo.json interchange format → a resolved `Diagram`.

The inverse of `to_kymojson.export` (KYMOJSON-MAP-001). A `.kymo.json` file already
holds a fully-resolved model (positions baked in), so `cli` renders it directly with
no layout / alignment pass — exactly like a `.bpmn` import. JSON arrays are coerced
back to the tuple-typed model fields so a loaded `Diagram` is identical to a
natively-parsed one. Unknown top-level fields are ignored (forward compatibility).
"""
from __future__ import annotations

import json

from .model import Component, Diagram, Edge, Region

FORMAT = "kymo.json"


def _pt(v):
    """JSON array → tuple for a point/bounds field; pass through None / scalars."""
    return tuple(v) if isinstance(v, list) else v


def _pts(v):
    """JSON array-of-points → list of tuples; pass through None."""
    return [tuple(p) for p in v] if isinstance(v, list) else v


def _component(d: dict) -> Component:
    return Component(
        id=d["id"], name=d["name"], subtitle=d["subtitle"], icon=d["icon"],
        shape=d["shape"], accent=d["accent"], pos=_pt(d["pos"]), size=_pt(d["size"]),
        parent=d["parent"], align=d["align"], align_gap=d["align_gap"],
        align_offset=_pt(d["align_offset"]), label_box=_pt(d.get("label_box")),
    )


def _region(d: dict) -> Region:
    return Region(
        id=d["id"], label=d["label"], bounds=_pt(d["bounds"]), contains=list(d["contains"]),
        padding=_pt(d["padding"]), padding_bottom=d["padding_bottom"], style=d["style"],
        icon=d["icon"], layout=d["layout"], pos=_pt(d["pos"]), gap=d["gap"],
        align=d["align"], visible=d["visible"], border_dash=_pt(d["border_dash"]),
        border_stroke=d["border_stroke"], label_anchor=d["label_anchor"],
        label_position=d["label_position"],
    )


def _edge(d: dict) -> Edge:
    return Edge(
        src=d["src"], dst=d["dst"], label=d["label"], style=d["style"],
        src_anchor=d["src_anchor"], dst_anchor=d["dst_anchor"], route=d["route"],
        via=_pts(d["via"]), src_offset=_pt(d["src_offset"]), dst_offset=_pt(d["dst_offset"]),
        label_offset=_pt(d["label_offset"]), label_anchor=d["label_anchor"],
        label_small=d["label_small"], label_pos=_pt(d["label_pos"]), dashed=d["dashed"],
        no_arrow=d["no_arrow"], trunk_offset=d["trunk_offset"], shared_port=d["shared_port"],
        points=_pts(d["points"]), bpmn_flow=d["bpmn_flow"],
    )


def _layout_node(n: dict):
    """Canonical layout-tree node → native tuple (`("id", cid)` / `("group", dir,
    [children])`); the group's children stay a mutable list (the layout pass mutates it)."""
    if n["t"] == "id":
        return ("id", n["id"])
    return ("group", n["dir"], [_layout_node(c) for c in n["children"]])


def model_from_dict(d: dict) -> Diagram:
    """Build a `Diagram` from a canonical model **body** dict (the `diagram` payload,
    i.e. what `to_kymojson.model_dict` emits and the Rust core returns)."""
    return Diagram(
        width=d["width"], height=d["height"], title=d["title"], subtitle=d["subtitle"],
        components=[_component(c) for c in d["components"]],
        regions=[_region(r) for r in d["regions"]],
        edges=[_edge(e) for e in d["edges"]],
        layout_trees=[_layout_node(t) for t in d.get("layout_trees", [])],
    )


def parse(text: str) -> Diagram:
    """Load a `.kymo.json` string (versioned envelope) into a fully-resolved `Diagram`."""
    payload = json.loads(text)
    fmt = payload.get("format")
    if fmt != FORMAT:
        raise ValueError(f"not a kymo.json document (format={fmt!r}, expected {FORMAT!r})")
    return model_from_dict(payload["diagram"])
